### PR TITLE
Provide a basic interface for updating metadata

### PR DESCRIPTION
**Pull request by kaojohnny**
connects #877

### CODE REVIEW

File: pkg/core/utils/utils.go [hunk 1]
```diff
@@ -63,3 +67,26 @@ func strAt(slice []string, str string) int {
 	}
 	return -1
 }
+
+func SliceContains(in interface{}, elem interface{}) bool {
+	inValue := reflect.ValueOf(in)
+	inType := inValue.Type()
+
+	if inType.Kind() == reflect.Slice {
```
Comment (Steven-Chan) [inline pkg/core/utils/utils.go]: Using reflect here looks overkill, the function name has already suggested that the input is a slice.

File: pkg/auth/dependency/provider/password/mock_password.go [hunk 2]
```diff
@@ -40,6 +41,19 @@ func (m *MockProvider) IsAuthDataMatching(authData map[string]string) bool {
 	return m.authDataChecker.isMatching(authData)
 }
 
+func (m *MockProvider) GetLoginIDMetadataFlattenedKeys() []string {
+	var output []string
+	for _, keys := range m.loginIDMetadataKeys {
+		for _, key := range keys {
+			if !utils.SliceContains(output, key) {
```
Comment (Steven-Chan) [inline pkg/auth/dependency/provider/password/mock_password.go]: This ends up with a 3-level nested loop, which the last layer is not needed.

Something like this: https://kylewbanks.com/blog/creating-unique-slices-in-go

File: pkg/auth/handler/update_metadata.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	inMetadata map[string]interface{},
+) map[string]interface{} {
+	loginIDMetadataKeys := h.PasswordAuthProvider.GetLoginIDMetadataFlattenedKeys()
+
+	output := make(map[string]interface{})
+	for k := range orgMetadata {
+		output[k] = orgMetadata[k]
+	}
+	for k, v := range inMetadata {
+		if !utils.SliceContains(loginIDMetadataKeys, k) {
```
[the remark this answers is not in the document]
Comment (kaojohnny) [reply pkg/auth/handler/update_metadata.go]: This PR is mainly for providing very basic functionality for update metadata for JS SDK, after https://github.com/SkygearIO/features/pull/295 is confirmed, these codes should be removed afterward.

[the remark this answers is not in the document]
Comment (Steven-Chan) [reply pkg/auth/handler/update_metadata.go]: Got it, i suggest you add comment to this.

**Comment by kaojohnny** [comment]
@Steven-Chan I don't think we should support update someone else's metadata, do you think it is okay to change `skygear.auth.updateMetadata(user)` to `skygear.auth.updateCurrentUserMetadata()`?

File: pkg/core/utils/utils.go [hunk 4]
```diff
@@ -63,3 +67,23 @@ func strAt(slice []string, str string) int {
 	}
 	return -1
 }
+
+func SliceContains(in interface{}, elem interface{}) bool {
+	inValue := reflect.ValueOf(in)
```
Comment (Steven-Chan) [inline pkg/core/utils/utils.go]: > Using reflect here looks overkill, the function name has already suggested that the input is a slice.

I didn't look clearly enough to realise the whole function is using reflection. I was trying to say using reflection for this function looks overkill. 

We don't actually have a case that needs to find a `interface{}` in a `[]interface{}`, so having a `StringSliceContains` is good enough for this case. This is where golang suffers when we want to use generic.

**Comment by Steven-Chan** [comment]
> @Steven-Chan I don't think we should support update someone else's metadata, do you think it is okay to change `skygear.auth.updateMetadata(user)` to `skygear.auth.updateCurrentUserMetadata()`?

IMO, there must be a way for the developer to update a specific user's metadata, e.g. user management for admin. And if we don't open up the db to the developers, we will need to provide an API to do it.

I was thinking `skygear.auth.updateMetadata({ metadata: user.metadata })` may be better in the example.

**Comment by chpapa** [comment]
> IMO, there must be a way for the developer to update a specific user's metadata, e.g. user management for admin. And if we don't open up the db to the developers, we will need to provide an API to do it.
> 
> I was thinking `skygear.auth.updateMetadata({ metadata: user.metadata })` may be better in the example.

Probably as an admin only functions?

File: pkg/auth/handler/update_metadata.go [hunk 5]
```diff
@@ -125,11 +125,11 @@ func (h UpdateMetadataHandler) mergeMetadata(
...
 	output := make(map[string]interface{})
-	for k := range orgMetadata {
-		output[k] = orgMetadata[k]
-	}
 	for k, v := range inMetadata {
-		if !utils.StringSliceContains(loginIDMetadataKeys, k) {
+		output[k] = v
+	}
+	for k, v := range orgMetadata {
+		if utils.StringSliceContains(loginIDMetadataKeys, k) {
```
Comment (Steven-Chan) [inline pkg/auth/handler/update_metadata.go]: Looks weird, why update with old data after new data?

I thought the previous implementation is able to delete attribute by setting value to `null`? This is the same behaviour as v1 record save. I think you may write a more complete test case, deleting non login id keys.

Comment (kaojohnny) [reply pkg/auth/handler/update_metadata.go]: My original intention is to copy all attributes from new data and overwrite the original loginID value.

I update the implementation as:

```
	output := make(map[string]interface{})
	// copy loginID to the profile
	for _, v := range loginIDMetadataKeys {
		if pv, ok := orgMetadata[v]; ok {
			output[v] = pv
		}
	}
	// copy rest attributes to the profile
	for k, v := range inMetadata {
		if !utils.StringSliceContains(loginIDMetadataKeys, k) {
			output[k] = v
		}
	}
```

Feel free to Let me know your comments about this. Thanks.

**Comment by kaojohnny** [comment]
> IMO, there must be a way for the developer to update a specific user's metadata, e.g. user management for admin. And if we don't open up the db to the developers, we will need to provide an API to do it.

> I was thinking skygear.auth.updateMetadata({ metadata: user.metadata }) may be better in the example.

>> Probably as an admin only functions?

Probably we can adjust the payload of `/update_metadata` to 

```
{
  user_id: <user_id>,
  metadata: <metadata>
}
```

1. In JS SDK, the API is `updateMetadata(metadata: user.metadata)`, `user_id` is current user id.
2. From cloud function, a developer can set `user_id` and assign master key to grant to update specific user's metadata.

**Pull request merged**